Update to 0.8.0 (#416)
ni/nidaqmx-python: CHANGELOG.md

All notable changes to this project will be documented in this file.

## 0.8.0

* ### Merged Pull Requests

>     * [Query: Closed PRs with the label: interpreter_implementation](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_implementation+is%3Aclosed)

Reviewer: umm, these link to issues, not PRs

Reviewer: fixed

    * [Query: Closed PRs with the label: library_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: grpc_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: test_improvements to increase the test coverage](https://github.com/ni/nidaqmx-python/issues?q=label%3Atest_improvements+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_fixes to fix all the issues in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_fixes+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_testcase_update to make the existing test cases run with both the interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_testcase_updates+is%3Aclosed)

>     * [Query: Closed PRs with the label: event_handling that takes care of the event handling in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Aevent_handling+is%3Aclosed)

Reviewer: deleted

* ### Major Changes

    * Added support for communicating with DAQmx devices through gRPC using [NI gRPC Device Server](https://github.com/ni/grpc-device). This enables using DAQmx with the MeasurementLink session management service.
        * The initialization methods for `Task`, `Scale`, and other classes now accept a keyword-only `grpc_options` parameter. Pass a `GrpcSessionsOptions` object to enable gRPC support.
        * The `System` class now has a `remote()` method which accepts a `GrpcSessionOptions` object.

>         * The signature of the collection classes' __init__ methods have been changed and any code that constructs these objects directly should be updated to get them from the appropriate object (e.g. System.devices, Task.channels)

Reviewer: Oops, this is redundant with "The initialization methods... now accept a keyword-only grpc_options parameter"

Reviewer: I deleted this bullet and added a bullet to internal details about how internal initialization methods take an interpreter parameter.

All notable changes to this project will be documented in this file.

## 0.8.0

* ### Merged Pull Requests
    * [Full changelog: 0.7.0...0.8.0](https://github.com/ni/nidaqmx-python/compare/0.7.0...0.8.0)
    * [Query: Closed PRs with the label: interpreter_implementation](https://github.com/ni/nidaqmx-python/pulls?q=label%3Ainterpreter_implementation+is%3Aclosed)
    * [Query: Closed PRs with the label: library_interpreter](https://github.com/ni/nidaqmx-python/pulls?q=label%3Alibrary_interpreter+is%3Aclosed)

>     * [Query: Closed PRs with the label: grpc_interpreter](https://github.com/ni/nidaqmx-python/pulls?q=label%3Alibrary_interpreter+is%3Aclosed)

Reviewer: wrong link

    * [NI gRPC Device Server](https://github.com/ni/grpc-device) version 2.2 or later is required. Older versions of NI gRPC Device Server are unsupported because they are missing bug fixes needed to support nidaqmx-python.
    * The following functions will now emit `DeprecationWarning` when called:
        * `nidaqmx.errors` module: `check_for_error`, `is_string_buffer_too_small`, and `is_array_buffer_too_small` are `ctypes`-specific helper functions, so they have been moved to an internal module.
        * `System` class: `set_analog_power_up_states` does not support `PowerUpStates.TRISTATE` and `get_analog_power_up_states` has design issues that make the previous implementation unworkable, so they have been deprecated in favor of `set_analog_power_up_states_with_output_type` and `get_analog_power_up_states_with_output_type`.
    * The internals of the `nidaqmx` package have been refactored to support this change:
        * Access to the DAQmx driver is now handled by the internal `BaseInterpreter` abstract base class. There are separate implementations of this abstract base class for `ctypes` vs. gRPC.

>         * Added support for communicating with DAQmx devices through gRPC using [NI gRPC Device Server](https://github.com/ni/grpc-device). This enables using DAQmx with the MeasurementLink session management service.

Reviewer: Why is this down here? This should be the top bullet

Reviewer: fixed

Reviewer 2: Sorry, I think you meant only the stub generation related things to internals from your previous comment, I misunderstood that for this as well and moved.

>         * For communicating with the device through gRPC, the following changes have been made:

Reviewer: This is redundant with "The internals have been refactored to support this change"

Reviewer: deleted

All notable changes to this project will be documented in this file.

> ## 0.7.2

Reviewer: The next release will be 0.8.0

    * [Query: Closed PRs with the label: library_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: grpc_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: test_improvements to increase the test coverage](https://github.com/ni/nidaqmx-python/issues?q=label%3Atest_improvements+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_fixes to fix all the issues in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_fixes+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_testcase_update to make the existing test cases run with both the interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_testcase_updates+is%3Aclosed)
    * [Query: Closed PRs with the label: event_handling that takes care of the event handling in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Aevent_handling+is%3Aclosed)
* ### Resolved Issues
    * ...
* ### Major Changes

>     * Introduced library and gRPC intepreters to communicate with the device, either directly or through gRPC respectively.

Reviewer 2: ```suggestion
    * Introduced library and gRPC interpreters to communicate with the device, either directly or through gRPC respectively.
```

>     * Introduced library and gRPC intepreters to communicate with the device, either directly or through gRPC respectively.

Reviewer: Use user-facing terminology. "interpreter" is an internal implementation detail.

    * The `library_interpreter` will contain all the c level implementation of various functions that can be used to directly communicate with the instrument
    * For the implemenation of `grpc_interpreter` the following changes have been made:
        * Added a stub generator which will generate the stubs based on the proto files present in `src/codegen/protos`
        and the files will be generate into `generator/nidaqmx/_stubs`.
        * Updated the existing generator to generate the `grpc_interpreter` based on the metadata provided.
        * Created `grpc_session_options` which defines the necessary information that is required to communicate via gRPC.
    * Updated task, system and scale objects to taken in `grpc_session_options` as an optional parameter and based on this  information, the interpreter to be used will be decided.
    * Updated the existing tests to run using both `library_interpreter` and `grpc_interpreter`.
    * Added multiple new test cases to improve the test coverage.

> ## 0.7.1

Reviewer: This is a placeholder for the next release.

Reviewer: 0.8.0 is the next release

Reviewer: This is content for the next release, which we are now planning to call 0.8.0. Delete the 0.8.1 section and merge its content into 0.8.0 or delete its content if it's already covered.

    * [Query: Closed PRs with the label: library_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: grpc_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: test_improvements to increase the test coverage](https://github.com/ni/nidaqmx-python/issues?q=label%3Atest_improvements+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_fixes to fix all the issues in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_fixes+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_testcase_update to make the existing test cases run with both the interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_testcase_updates+is%3Aclosed)
    * [Query: Closed PRs with the label: event_handling that takes care of the event handling in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Aevent_handling+is%3Aclosed)
* ### Resolved Issues

>     * ...

Reviewer 3: You can just delete this

* ### Major Changes

>     * Added support to communicate with the DAQmx devices through gRPC.

Reviewer 3: Add a link to grpc-device?

```suggestion
    * Added support to communicate with the DAQmx devices through gRPC using [gRPC Device Server](https://github.com/ni/grpc-device).
```

    * For communicating with the device through gRPC, the following changes have been made:

>         * Added a stub generator which will generate the gRPC stub files based on the proto files present in `src/codegen/protos`
>         and the files will be generate into `generator/nidaqmx/_stubs`.

Reviewer 3: Small grammar recommendation. I don't think the `and` really follows, and updated to `generated`. I also removed the line-break since we don't seem to be doing that anywhere else here.

```suggestion
        * Added a stub generator which will generate the gRPC stub files based on the proto files present in `src/codegen/protos`. The files will be generated into `generator/nidaqmx/_stubs`.
```

>         * Updated the existing generator to generate the `grpc_interpreter` based on the metadata provided, this will implement all the necessary methods using the gRPC stubs.

Reviewer 3: Either two sentences or a semicolon or maybe this is more succinct...

```suggestion
        * Updated the existing generator to generate the `grpc_interpreter` based on the metadata provided that will implement all the necessary methods using the gRPC stubs.
```

        * Created `grpc_session_options` which defines the necessary information that is required to communicate via gRPC.

>     * Updated task, system and scale objects to taken in `grpc_session_options` as an optional parameter and based on this information, the interpreter to be used will be decided.

Reviewer: What other internals did we deprecate or remove? Some functions in `nidaqmx.errors`

    * Updated the existing tests to run with and without gRPC support.
    * Added multiple new test cases to improve the test coverage.
    * Currently, the behavior of all the operations will be similar when communicating with and without gRPC with the following exceptions:

>         * When reading values from a `read_stream`, the initial values sent by the user will not retained in the read array. Based on the number of samples read, the unread samples will be set to default value of the datatype specified.

Reviewer: This belongs in a "known issues" list. 

Are there any other known issues? (excluding tech debt)

>     * In order to make the behavior of each opration similar when communicating with the device with or without gRPC, the grpc device server has been updated with the appropriate changes. In order for these fixes to work, gRPC device version of 2.2 or higher has to be used.

Reviewer 3: typo and links

```suggestion
    * In order to make the behavior of each operation similar when communicating with the device with or without gRPC, the [gRPC Device Server](https://github.com/ni/grpc-device) has been updated with the appropriate changes. In order for these fixes to work, gRPC Device Server version of 2.2 or higher has to be used.
```

Reviewer: ```
* NI gRPC Device Server version 2.2 or later is required. Older versions of NI gRPC Device Server are unsupported because they are missing bug fixes needed to support nidaqmx-python.
```

    * [Query: Closed PRs with the label: library_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: grpc_interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Alibrary_interpreter+is%3Aclosed)
    * [Query: Closed PRs with the label: test_improvements to increase the test coverage](https://github.com/ni/nidaqmx-python/issues?q=label%3Atest_improvements+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_fixes to fix all the issues in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_fixes+is%3Aclosed)
    * [Query: Closed PRs with the label: interpreter_testcase_update to make the existing test cases run with both the interpreter](https://github.com/ni/nidaqmx-python/issues?q=label%3Ainterpreter_testcase_updates+is%3Aclosed)
    * [Query: Closed PRs with the label: event_handling that takes care of the event handling in the interpreters](https://github.com/ni/nidaqmx-python/issues?q=label%3Aevent_handling+is%3Aclosed)

* ### Major Changes

>     * Added support for communicating with DAQmx devices through gRPC using [NI gRPC Device Server](https://github.com/ni/grpc-device). This enables the users to use DAQmx session drivers with measurement link session manager.

Reviewer: definite article, branding

```suggestion
    * Added support for communicating with DAQmx devices through gRPC using [NI gRPC Device Server](https://github.com/ni/grpc-device). This enables using DAQmx with the MeasurementLink session management service.
```

reference for branding: https://www.ni.com/docs/en-US/bundle/measurementlink/page/session-management.html

Reviewer: Also move the 2nd sentence to the 1st bullet

    * For communicating with the device through gRPC, the following changes have been made:

>         * Added a stub generator which will generate the gRPC stub files based on the proto files present in `src/codegen/protos`. The files will be generated into `generator/nidaqmx/_stubs`.

Reviewer: Move to internals

        * The initialization methods for `Task`, `Scale`, and other classes now accept a keyword-only `grpc_options` parameter. Pass a `GrpcSessionsOptions` object to enable gRPC support.
        * The `System` class now has a `remote()` method which accepts a `GrpcSessionOptions` object.
        * The internals of the `nidaqmx` package have been refactored to support this change:
            * Access to the DAQmx driver is now handled by the internal `BaseInterpreter` abstract base class. There are separate implementations of this abstract base class for `ctypes` vs. gRPC.
            * The internal `nidaqmx._task_modules.read_functions` and `nidaqmx._task_modules.write_functions` modules have been removed. If your application uses these modules, you must update it to use public APIs such as `task.read()`/`task.write()`, `task.in_stream.read()`/`task.out_stream.write()`, or `nidaqmx.stream_readers`/`nidaqmx.stream_writers`.

>         * The following functions have been deprecated as they have been either migrated to a different module or the support has been terminated. These functions are still usable, but will emit a `DeprecationWarning` on usage:
>             * `check_for_error`, `is_string_buffer_too_small`, `is_array_buffer_too_small` in `errors.py`
>             * `set_analog_power_up_states` and `get_analog_power_up_states` in `system.py`.

Reviewer: Move out of gRPC section (unindent)

```suggestion
        * The following functions will now emit `DeprecationWarning` when called:
            * `nidaqmx.errors`: `check_for_error`, `is_string_buffer_too_small`, and `is_array_buffer_too_small` are `ctypes`-specific helper functions, so they have been moved to an internal module.
            * `System` class: `set_analog_power_up_states` does not support `PowerUpStates.TRISTATE` and `get_analog_power_up_states` has design issues that make the previous implementation unworkable, so they have been deprecated in favor of `set_analog_power_up_states_with_output_type` and `get_analog_power_up_states_with_output_type`.
```

    * Updated the existing tests to run with and without gRPC support.

>     * Added multiple test cases to improve the overall test coverage.

Reviewer: Move to internals

>     * [NI gRPC Device Server](https://github.com/ni/grpc-device) version 2.2 or later is required. Older versions of NI gRPC Device Server are unsupported because they are missing the bug fixes needed to support nidaqmx-python.

Reviewer: Move above internals

Definite article is not necessary, it's okay to have an indefinite number of bug fixes

```suggestion
    * [NI gRPC Device Server](https://github.com/ni/grpc-device) version 2.2 or later is required. Older versions of NI gRPC Device Server are unsupported because they are missing bug fixes needed to support nidaqmx-python.
```

* ### Known Issues

>    * When reading an array of values from any `ReadStream`, the initial array values sent by the user will not be retained. Based on the number of samples read, the unread samples will be set to the default value of the datatype specified.

Reviewer: There is no such class as `ReadStream`.

Based on the investigation I did yesterday, I don't think we should list this as a known issue, at least in its current form.

Reviewer 2: can we mention it as `nidaqmx.stream_readers` classes?

Reviewer: Based on the investigation I did Thursday, I don't think we should list this as a known issue.

Reviewer: deleted

>    * When comparing multiple DAQmx objects (like task, device,etc..) which are created using gRPC with one another, the comparison will be improper. 

Reviewer: ```suggestion
   * Comparisons between DAQmx object instances do not take gRPC remoting into account. For example, `Device("Dev1")` refers to a local device and `Device("Dev1", grpc_options=...)` refers to a remote device, but they are considered equal. Likewise, two instances of `Device("Dev1", grpc_options=...)` with different remote hosts are considered equal.
 ```
